source/extensions/transport_sockets/tls/ssl_handshaker.cc

```diff
@@ -44,14 +44,16 @@ Envoy::Ssl::ClientValidationStatus SslExtendedSocketInfoImpl::certificateValidat
   return certificate_validation_status_;
 }
 
-void SslExtendedSocketInfoImpl::onCertificateValidationCompleted(bool succeeded) {
+void SslExtendedSocketInfoImpl::onCertificateValidationCompleted(bool succeeded, bool async) {
   cert_validation_result_ =
       succeeded ? Ssl::ValidateStatus::Successful : Ssl::ValidateStatus::Failed;
   if (cert_validate_result_callback_.has_value()) {
     // This is an async cert validation.
     cert_validate_result_callback_.reset();
```

Comment:
We should change this interface
https://github.com/envoyproxy/envoy/blob/a0c50af79c7790227294f8e3a2523fd2b6255b67/envoy/ssl/ssl_socket_extended_info.h#L62-L66

The callback needn't to be created on the fly, we can create the callback in the constructor of SslExtendedSocketInfoImpl. Then we change `createValidateResultCallback` interface to `getValidateResultCallback`

Reply:
Actually, `createValidateResultCallback` will also change cert validation state in `SslExtendedSocketInfoImpl`. So even if we create the validate result callback in the constructor, we still need an non-const interface to notice SSL extended socket info that the validation has been started. I suggest a comment change only.

Reply:
Yes, that's true, I'm ok with only comment change.

Reply:
Unmatched comments removed.